Spigot-Server-Patches/0415-Cache-block-data-strings.patch

```diff
@@ -0,0 +1,64 @@
+From 842c4b4a50176ecf3503f7b82a05fc1fbffb368e Mon Sep 17 00:00:00 2001
+From: miclebrick <miclebrick@outlook.com>
+Date: Thu, 6 Dec 2018 19:52:50 -0500
+Subject: [PATCH] Cache block data strings
+
+
+diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
+index 9c8c3304..4af9a168 100644
+--- a/src/main/java/net/minecraft/server/MinecraftServer.java
++++ b/src/main/java/net/minecraft/server/MinecraftServer.java
+@@ -1743,6 +1743,7 @@ public abstract class MinecraftServer implements IAsyncTaskHandler, IMojangStati
+             this.resourcePackRepository.a();
+             this.a(this.getWorldServer(DimensionManager.OVERWORLD).getWorldData());
+             this.getPlayerList().reload();
++            org.bukkit.craftbukkit.block.data.CraftBlockData.reloadCache(); // Paper - cache block data strings, they can be defined by datapacks so refresh it here
+         }
+     }
+ 
+diff --git a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+index b9279d92..076f8938 100644
+--- a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
++++ b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+@@ -468,9 +468,38 @@ public class CraftBlockData implements BlockData {
+         Preconditions.checkState(MAP.put(nms, bukkit) == null, "Duplicate mapping %s->%s", nms, bukkit);
+     }
+ 
++    // Paper start - cache block data strings
++    private static Map<String, CraftBlockData> stringDataCache = new HashMap<>();
++
++    static {
++        // cache all of the default states at startup, will not cache ones with the custom states inside of the
++        // brackets in a different order, though
++        reloadCache();
++    }
++
++    public static void reloadCache() {
++        stringDataCache.clear();
++        Block.REGISTRY_ID.forEach(blockData -> stringDataCache.put(blockData.toString(), blockData.createCraftBlockData()));
++    }
++
+     public static CraftBlockData newData(Material material, String data) {
+         Preconditions.checkArgument(material == null || material.isBlock(), "Cannot get data for not block %s", material);
+ 
++        if (material != null) {
```

Comment:
should paper start here and paper end above, i feel its not intuitive these blocks go together at immediate glance.

but otherwise, if @PaperMC/development-team can rebase and this still applies good (please reorder patch to end) this seems fine.